nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/kafka/processors/producer/convert/RecordWrapperStreamKafkaRecordConverter.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.nifi.serialization.RecordSetWriterFactory;
+import org.apache.nifi.serialization.record.MapRecord;
+import org.apache.nifi.serialization.record.PushBackRecordSet;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordSet;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * {@link KafkaRecordConverter} implementation for transforming NiFi
+ * {@link org.apache.nifi.serialization.record.Record} objects to {@link KafkaRecord} for publish to
+ * Kafka.
+ */
+public class RecordWrapperStreamKafkaRecordConverter implements KafkaRecordConverter {
+    private final FlowFile flowFile;
+    private final RecordMetadataStrategy metadataStrategy;
+    private final RecordReaderFactory readerFactory;
+    private final RecordSetWriterFactory writerFactory;
+    private final RecordSetWriterFactory keyWriterFactory;
+    private final int maxMessageSize;
+    private final ComponentLog logger;
+
+    public RecordWrapperStreamKafkaRecordConverter(
+            final FlowFile flowFile,
+            final RecordMetadataStrategy metadataStrategy,
+            final RecordReaderFactory readerFactory,
+            final RecordSetWriterFactory writerFactory,
+            final RecordSetWriterFactory keyWriterFactory,
+            final int maxMessageSize,
+            final ComponentLog logger) {
+        this.flowFile = flowFile;
+        this.metadataStrategy = metadataStrategy;
+        this.readerFactory = readerFactory;
+        this.writerFactory = writerFactory;
+        this.keyWriterFactory = keyWriterFactory;
+        this.maxMessageSize = maxMessageSize;
+        this.logger = logger;
+    }
+
+    @Override
+    public Iterator<KafkaRecord> convert(
+            final Map<String, String> attributes, final InputStream in, final long inputLength)
+            throws IOException {
+        try {
+            final RecordReader reader = readerFactory.createRecordReader(attributes, in, inputLength, logger);
+            final RecordSet recordSet = reader.createRecordSet();
+            final PushBackRecordSet pushBackRecordSet = new PushBackRecordSet(recordSet);
+            return toKafkaRecordIterator(pushBackRecordSet);
+        } catch (MalformedRecordException | SchemaNotFoundException e) {
+            throw new IOException("Stream to Record conversion failed", e);
+        }
+    }
+
+    private Iterator<KafkaRecord> toKafkaRecordIterator(
+            final PushBackRecordSet pushBackRecordSet) {
+        return new Iterator<>() {
+            @Override
+            public boolean hasNext() {
+                try {
+                    return pushBackRecordSet.isAnotherRecord();
+                } catch (final IOException e) {
+                    throw new UncheckedIOException(e);
```

Comment:
A message should be included:
```suggestion
                    throw new UncheckedIOException("Record evaluation failed", e);
```